File: src/main/java/org/jukeboxmc/block/behavior/BlockComposter.java
package org.jukeboxmc.block.behavior;

import org.cloudburstmc.nbt.NbtMap;
import org.jukeboxmc.block.Block;
import org.jukeboxmc.util.Identifier;

/**
 * @author LucGamesYT
 * @version 1.0
 */
public class BlockComposter extends Block {

    public BlockComposter( Identifier identifier ) {
        super( identifier );
    }

    public BlockComposter( Identifier identifier, NbtMap blockStates ) {
        super( identifier, blockStates );
    }

    public void setComposterFillLevel( int value ) {
        this.setState( "composter_fill_level", value );
    }

    public int getComposterFillLevel() {
        return this.stateExists( "composter_fill_level" ) ? this.getIntState( "composter_fill_level" ) : 0;
    }
}

File: JukeboxMC-Server/src/main/kotlin/org/jukeboxmc/server/block/behavior/BlockComposter.kt
package org.jukeboxmc.server.block.behavior

import org.cloudburstmc.nbt.NbtMap
import org.jukeboxmc.api.Identifier
import org.jukeboxmc.api.block.Composter
import org.jukeboxmc.api.block.data.BlockFace
import org.jukeboxmc.api.math.Vector
import org.jukeboxmc.server.block.JukeboxBlock
import org.jukeboxmc.server.item.JukeboxItem
import org.jukeboxmc.server.player.JukeboxPlayer
import org.jukeboxmc.server.world.JukeboxWorld

class BlockComposter(identifier: Identifier, blockStates: NbtMap?) : JukeboxBlock(identifier, blockStates),
    Composter {

    override fun placeBlock(
        player: JukeboxPlayer,
        world: JukeboxWorld,
        blockPosition: Vector,
        placePosition: Vector,
        clickedPosition: Vector,
        itemInHand: JukeboxItem,
        blockFace: BlockFace
    ): Boolean {
        val block = world.getBlock(placePosition)
        if (block is BlockWater && block.getLiquidDepth() == 0) {
            world.setBlock(placePosition, block, 1, false)
        }
        return super.placeBlock(player, world, blockPosition, placePosition, clickedPosition, itemInHand, blockFace)
    }

    override fun getComposterFillLevel(): Int {
        return this.getIntState("composter_fill_level")
    }

    override fun setComposterFillLevel(value: Int): Composter {
        return this.setState("composter_fill_level", value)
    }

    override fun getWaterLoggingLevel(): Int {
        return 1
    }
}